cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaBrokerConfigurationDiff.java

```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+
+package io.strimzi.operator.cluster.operator.resource;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import io.fabric8.zjsonpatch.JsonDiff;
+import io.strimzi.kafka.config.model.ConfigModel;
+import io.strimzi.kafka.config.model.Scope;
+import io.strimzi.operator.cluster.model.KafkaConfiguration;
+import io.strimzi.operator.cluster.model.KafkaVersion;
+import io.strimzi.operator.cluster.model.ModelUtils;
+import io.strimzi.operator.common.operator.resource.AbstractResourceDiff;
+import org.apache.kafka.clients.admin.AlterConfigOp;
+import org.apache.kafka.clients.admin.Config;
+import org.apache.kafka.clients.admin.ConfigEntry;
+import org.apache.kafka.common.config.ConfigResource;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static io.fabric8.kubernetes.client.internal.PatchUtils.patchMapper;
+
+/**
+ * Computes a diff between the current config (supplied as a Map ConfigResource) and the desired config (supplied as a String).
+ * An algorithm:
+ *  1. Create map from supplied desired String
+ *  2. Fill placeholders (e.g. ${BROKER_ID}) in desired map
+ *  3a. Loop over all entries. If the entry is in IGNORABLE_PROPERTIES or entry.value from desired is equal to entry.value from current, do nothing
+ *      else add it to the diff
+ *  3b. If entry was removed from desired, add it to the diff with null value.
+ *  3c. If custom entry was removed, delete property
+ *
+ */
+public class KafkaBrokerConfigurationDiff extends AbstractResourceDiff {
```

Comment:
This class presents an API for callers which allows them to do all dangerous things, like calling `computeDiff()` for themselves. Having public methods which mutate the instance makes it much harder to reason about whether the code is correct, because I have to know whether any of those mutating methods is ever called. Using the API of the class to enforce immutability means I know, just from reading the classes code, that callers can't screw up the computed state. The API it _should_ present is that the constructor computes the diff and various things about that are exposed immutably to the caller via accessor methods.